Source/NSPathControl.m

```diff
@@ -0,0 +1,549 @@
+                                      assumeInside: YES];
+}
+
+- (instancetype) init
+{
+  self = [super init];
+  if (self != nil)
+    {
+      [self setPathStyle: NSPathStyleStandard];
+      [self setURL: nil];
+      [self setDelegate: nil];
+      [self setAllowedTypes: [NSArray arrayWithObject: NSFilenamesPboardType]];
+    }
+  return self;
+}
+
+- (void) dealloc
+{
+  [[self superview] removeTrackingRect: _trackingTag];
+  [super dealloc];
+}
+
+- (void) mouseEntered: (NSEvent *)event
+{
+  [_cell mouseEntered: event
+            withFrame: [self frame]
+               inView: self];
+}
+
+- (void) mouseExited: (NSEvent *)event
+{
+  [_cell mouseExited: event
+           withFrame: [self frame]
+              inView: self];
+}
+
+- (void) setPathStyle: (NSPathStyle)style
+{
+  [_cell setPathStyle: style];
+  [self setNeedsDisplay];
+}
+
+- (NSPathStyle) pathStyle
+{
+  return [_cell pathStyle];
+}
+
+- (NSPathComponentCell *) clickedPathComponentCell
+{
+  return [_cell clickedPathComponentCell];
+}
+
+- (NSArray *) pathComponentCells
+{
+  return [_cell pathComponentCells];
+}
+
+- (void) setPathComponentCells: (NSArray *)cells
+{
+  [_cell setPathComponentCells: cells];
+  [self setNeedsDisplay];
+}
+
+- (SEL) doubleAction;
+{
+  return [_cell doubleAction];
+}
+
+- (void) setDoubleAction: (SEL)doubleAction
+{
+  [_cell setDoubleAction: doubleAction];
+}
+
+- (void) _createPathItems
+{
+  NSArray *a = [_cell pathComponentCells];
+  NSEnumerator *en = [a objectEnumerator];
+  NSPathComponentCell *c = nil;
+  NSMutableArray *items = [NSMutableArray arrayWithCapacity: [a count]];
+  
+  while((c = [en nextObject]) != nil)
+    {
+      NSPathControlItem *pi = [c _pathControlItem];
+      [items addObject: pi];
+    }
+
+  [self setPathItems: [items copy]];
+}
+
+- (NSURL *) URL
+{
+  return [_cell URL];
+}
+
+- (void) setURL: (NSURL *)url
+{
+  [_cell setURL: url];
+  [self _createPathItems];
+  [self setNeedsDisplay];
+}
+
+- (id<NSPathControlDelegate>) delegate
+{
+  return _delegate;
+}
+
+- (void) setDelegate: (id<NSPathControlDelegate>) delegate
+{
+  _delegate = delegate;
+}
+
+- (NSDragOperation) draggingSourceOperationMaskForLocal: (BOOL)flag
+{
+  if (flag)
+    {
+      return _localMask;
+    }
+
+  return _remoteMask;
+}
+
+- (void) setDraggingSourceOperationMask: (NSDragOperation)mask 
+                               forLocal: (BOOL)local
+{
+  if (local)
+    {
+      _localMask = mask;
+    }
+  else
+    {
+      _remoteMask = mask;
+    }
+}
+
+- (NSArray *) allowedTypes;
+{
+  return [_cell allowedTypes];
+}
+
+- (void) setAllowedTypes: (NSArray *)allowedTypes
+{
+  [_cell setAllowedTypes: allowedTypes];
+  [self registerForDraggedTypes: allowedTypes];
+}
+
+- (NSPathControlItem *) clickedPathItem
+{
+  return [[self clickedPathComponentCell] _pathControlItem];
+}
+
+- (NSArray *) pathItems
+{
+  return _pathItems;
+}
+
+- (void) setPathItems: (NSArray *)items
+{
+  ASSIGNCOPY(_pathItems, items);
```

Comment:
If this gets called from the outside the cell and all its components will be out of sync.